### PR TITLE
Refactor UDFs and Bblfsh code

### PR DESCRIPTION
**Pull request by mcarmonaa**
Changes:

- All UDFs code has been modified to use them in an uniformly way. It tries to improve the 
understandability and maintainability of the code keeping a common interface and usage style. You can call an UDF just with `myCoolUDF(col1, col2, col3)(implicit SparkSession)`. The UDFs are `case objects` that extends `CustomUDF`. All of them implements `name`, `function` and `apply`. 

- All bblfsh code has been gathered under the `Bbflsh` object, so configuration and client is wrapped and used directly from here. It avoids to have `SparkSession`s, `Bblfsh.Config`s and `Bblfsh.Client` mixed with the rest of the engine code. To get configuration just `Bblfsh.getConfig`and for extractUast and filter nodes just use `Bblfsh.extractUAST` and `BBlfsh.filter`

- Bblfsh errors are logged too:
```scala
scala> engine.getRepositories.getHEAD.getCommits.getFirstReferenceCommit.getFiles.extractUASTs.show
17/11/14 13:14:37 WARN Bblfsh: FATAL app/src/main/res/values/strings.xml: unexpected error: runtime failure: missing driver for language "xml"
17/11/14 13:14:37 WARN Bblfsh: FATAL app/src/main/res/values/strings.xml: unexpected error: runtime failure: missing driver for language "xml"
17/11/14 13:14:37 WARN Bblfsh: FATAL settings.gradle: unexpected error: runtime failure: missing driver for language "gradle"
17/11/14 13:14:37 WARN Bblfsh: FATAL settings.gradle: unexpected error: runtime failure: missing driver for language "gradle"
17/11/14 13:14:37 WARN Bblfsh: FATAL settings.gradle: unexpected error: runtime failure: missing driver for language "gradle"
17/11/14 13:14:37 WARN Bblfsh: FATAL settings.gradle: unexpected error: runtime failure: missing driver for language "gradle"
17/11/14 13:14:37 WARN Bblfsh: FATAL app/src/main/res/values/dimens.xml: unexpected error: runtime failure: missing driver for language "xml"
17/11/14 13:14:37 WARN Bblfsh: FATAL app/src/main/res/values/dimens.xml: unexpected error: runtime failure: missing driver for language "xml"
17/11/14 13:14:37 WARN Bblfsh: FATAL app/proguard-rules.txt: unexpected error: runtime failure: missing driver for language "text"
17/11/14 13:14:37 WARN Bblfsh: FATAL app/proguard-rules.txt: unexpected error: runtime failure: missing driver for language "text"
17/11/14 13:14:37 WARN Bblfsh: FATAL Google Image Downloader/create_dir.py~: unexpected error: runtime failure: missing driver for language ""
17/11/14 13:14:37 WARN Bblfsh: FATAL Google Image Downloader/create_dir.py~: unexpected error: runtime failure: missing driver for language ""
17/11/14 13:14:37 WARN Bblfsh: FATAL app/src/main/res/menu/main.xml: unexpected error: runtime failure: missing driver for language "xml"
17/11/14 13:14:37 WARN Bblfsh: FATAL app/src/main/res/menu/main.xml: unexpected error: runtime failure: missing driver for language "xml"
17/11/14 13:14:37 WARN Bblfsh: FATAL app/.gitignore: unexpected error: runtime failure: missing driver for language ""
17/11/14 13:14:37 WARN Bblfsh: FATAL app/.gitignore: unexpected error: runtime failure: missing driver for language ""
17/11/14 13:14:37 WARN Bblfsh: FATAL app/src/main/res/menu/login.xml: unexpected error: runtime failure: missing driver for language "xml"
17/11/14 13:14:38 WARN Bblfsh: FATAL app/src/main/res/menu/login.xml: unexpected error: runtime failure: missing driver for language "xml"
17/11/14 13:14:38 WARN Bblfsh: FATAL gradlew.bat: unexpected error: runtime failure: missing driver for language "batchfile"
17/11/14 13:14:38 WARN Bblfsh: FATAL gradlew.bat: unexpected error: runtime failure: missing driver for language "batchfile"
```

### CODE REVIEW

File: src/main/scala/tech/sourced/engine/package.scala [hunk 1]
```diff
@@ -295,10 +294,9 @@ package object engine {
         throw new SparkException(s"DataFrame already contains a column named $outputColumn")
       }
 
-      df.withColumn(
-        outputColumn,
-        QueryXPathUDF(df.sparkSession, query)(df(queryColumn))
-      )
+      import org.apache.spark.sql.functions.lit
+      val queryB = df.sparkSession.sparkContext.broadcast(query)
```
Comment (ajnavarro) [inline src/main/scala/tech/sourced/engine/package.scala]: You don't need to broadcast values to UDF definitions

File: src/main/scala/tech/sourced/engine/package.scala [hunk 2]
```diff
@@ -57,10 +58,12 @@ package object engine {
     def registerUDFs(): Unit = {
       SessionFunctions.UDFtoRegister.foreach(customUDF => session.udf.register(
         customUDF.name,
-        customUDF.function(session)
+        customUDF.function
       ))
     }
 
+    Bblfsh.setConfig(session)
+    session.sparkContext.broadcast(Bblfsh.getConfig())
```
Comment (ajnavarro) [inline src/main/scala/tech/sourced/engine/package.scala]: That is doing something?

File: src/main/scala/tech/sourced/engine/udf/ExtractUASTsUDF.scala [hunk 3]
```diff
@@ -1,118 +1,76 @@
...
+  */
+object ExtractUASTsUDF {
+
+  def apply(path: Column,
+            content: Column,
+            lang: Column = null): Column = {
+    if (lang == null) {
+      ExtractUASTsWithoutLangUDF(path, content)
+    } else {
+      ExtractUASTsWithLangUDF.function(path, content, lang)
```
Comment (ajnavarro) [inline src/main/scala/tech/sourced/engine/udf/ExtractUASTsUDF.scala]: Why on `ExtractUASTsWithLangUDF` we need to call to function method and in `ExtractUASTsWithoutLangUDF` not?

File: src/main/scala/tech/sourced/engine/util/Bblfsh.scala [hunk 4]
```diff
@@ -17,34 +27,86 @@ object Bblfsh {
...
+    */
+  def extractUAST(path: String,
+                  content: Array[Byte],
+                  lang: String): Seq[Array[Byte]] = {
+
+    //FIXME(bzz): not everything is UTF-8 encoded :/
+
+    // if lang == null, it hasn't been classified yet
+    // so rely on bblfsh to guess this file's language
+    if (lang != null && !supportedLangs.contains(lang.toLowerCase())) {
```
Comment (ajnavarro) [inline src/main/scala/tech/sourced/engine/util/Bblfsh.scala]: supportedLangs are in lowercase too?

File: src/main/scala/tech/sourced/engine/util/Bblfsh.scala [hunk 5]
```diff
@@ -17,34 +27,86 @@ object Bblfsh {
...
+    } else {
+      config
+    }
+  }
+
+  private def getClient(): BblfshClient = synchronized {
     if (client == null) {
-      client = BblfshClient(config.host, config.port)
+      val Config(host, port) = getConfig()
+      client = BblfshClient(host, port)
```
Comment (ajnavarro) [inline src/main/scala/tech/sourced/engine/util/Bblfsh.scala]: This client will always have the same default configuration on remote workers. we need to broadcast configuration from the driver to the workers. I cannot see where we are doing this.

File: src/main/scala/tech/sourced/engine/package.scala (hunk 2 above)
Comment (mcarmonaa) [reply src/main/scala/tech/sourced/engine/package.scala]: Nope, let me move the bblfsh broadcasting configuration to the `Bblfsh` object

File: src/main/scala/tech/sourced/engine/package.scala (hunk 1 above)
Comment (mcarmonaa) [reply src/main/scala/tech/sourced/engine/package.scala]: We were doing it before in the [UDF wrapper object](https://github.com/src-d/engine/blob/master/src/main/scala/tech/sourced/engine/udf/QueryXPathUDF.scala#L20)

I'll remove it

File: src/main/scala/tech/sourced/engine/udf/ExtractUASTsUDF.scala (hunk 3 above)
Comment (mcarmonaa) [reply src/main/scala/tech/sourced/engine/udf/ExtractUASTsUDF.scala]: We don't need, I forgot changed it

File: src/main/scala/tech/sourced/engine/util/Bblfsh.scala (hunk 5 above)
Comment (mcarmonaa) [reply src/main/scala/tech/sourced/engine/util/Bblfsh.scala]: This is related with your first comment, let' modify this

File: src/main/scala/tech/sourced/engine/util/Bblfsh.scala (hunk 4 above)
Comment (mcarmonaa) [reply src/main/scala/tech/sourced/engine/util/Bblfsh.scala]: Yes, they are `val supportedLangs = Set("python", "java")`

Comment (ajnavarro) [reply src/main/scala/tech/sourced/engine/util/Bblfsh.scala]: Now thinking about this Set with the supported languages, Should we delete it? I mean, If bblfsh adds more languages and the user didn't precalculate languages using Enry, thew will not be able to send to bblfsh files with the new supported language.

**Comment by mcarmonaa** [comment]
@ajnavarro now the bblfsh configuration is broadcast from `object Engine` when `apply` is used.

In this way, bblfsh utils is agnostic of `SparkSessions` and `Broadcast` variables.

File: src/main/scala/tech/sourced/engine/Engine.scala [hunk 6]
```diff
@@ -182,7 +183,19 @@ object Engine {
...
   }
+
+  private def spreadBblfshConfig(session: SparkSession): Unit = {
+    val host = session.conf.get(bblfshHostKey, Bblfsh.defaultHost)
+    val port = session.conf.get(bblfshPortKey, Bblfsh.defaultPort.toString).toInt
+
+    val hostB = session.sparkContext.broadcast(host)
+    val portB = session.sparkContext.broadcast(port)
+
+    Bblfsh.setConfig(hostB.value, portB.value)
```
Comment (ajnavarro) [inline src/main/scala/tech/sourced/engine/Engine.scala]: This wont work. You are setting the configuration just to the local JVM. You should add as configuration the Broadcast value, not the value inside the broadcast itself.

File: src/main/scala/tech/sourced/engine/util/Bblfsh.scala (hunk 5 above)
Comment (mcarmonaa) [reply src/main/scala/tech/sourced/engine/util/Bblfsh.scala]: see [comment](https://github.com/src-d/engine/pull/187#issuecomment-344534977)

File: src/main/scala/tech/sourced/engine/util/Bblfsh.scala (hunk 4 above)
Comment (mcarmonaa) [reply src/main/scala/tech/sourced/engine/util/Bblfsh.scala]: see [comment](https://github.com/src-d/engine/pull/187#issuecomment-344534977)

Comment (erizocosmico) [reply src/main/scala/tech/sourced/engine/util/Bblfsh.scala]: This should be delegated to bblfsh imho

File: src/main/scala/tech/sourced/engine/Engine.scala (hunk 6 above)
Comment (mcarmonaa) [reply src/main/scala/tech/sourced/engine/Engine.scala]: Ok, so something like `Bblfsh.setConfig(hostB, portB)` and left the `Bblfsh.setConfig` implementation extract the values, right?

Comment (erizocosmico) [inline src/main/scala/tech/sourced/engine/Engine.scala]: Are you sure this works? Because even if the values of the host and port have been broadcasted, the Bblfsh object has not been mutated elsewhere

Comment (erizocosmico) [reply src/main/scala/tech/sourced/engine/Engine.scala]: Nope, that's mutating the local object as well. That's why before it was the way it was done

Comment (mcarmonaa) [reply src/main/scala/tech/sourced/engine/Engine.scala]: I see, thanks for the explanation @erizocosmico!

Because all of this, I wonder if it's actually needed broadcast the configuration then. I mean, the only purpose of this configuration is get a bblfsh client, and in the current implementation only one client is instantiated to reuse it purposes. So, isn't the configuration just sent once?

File: src/main/scala/tech/sourced/engine/util/Bblfsh.scala (hunk 4 above)
Comment (mcarmonaa) [reply src/main/scala/tech/sourced/engine/util/Bblfsh.scala]: We already had a Set to don't send `markdown` neither `text` as optimization. Should I come back to this approach? Or just not exclude any language?

File: src/main/scala/tech/sourced/engine/Engine.scala (hunk 6 above)
Comment (erizocosmico) [reply src/main/scala/tech/sourced/engine/Engine.scala]: But you don't have the client instantiated in other nodes. So yes, you need to send it.

If you do a setConfig in node A, node B still has a `null` `client`, they run on completely different JVMs, AFAIK.

In local mode, of course, this works, because they share the same JVM.

File: src/main/scala/tech/sourced/engine/util/Bblfsh.scala (hunk 4 above)
Comment (erizocosmico) [reply src/main/scala/tech/sourced/engine/util/Bblfsh.scala]: Things we know will never have a driver, such as markdown and text, make sense to optimize. But languages that might have it in the future it's better to delegate to bblfsh, otherwise we must maintain that list and people using engine of version X will need to upgrade if they want to use language Y that was not supported back then.

File: src/main/scala/tech/sourced/engine/Engine.scala (hunk 6 above)
Comment (mcarmonaa) [reply src/main/scala/tech/sourced/engine/Engine.scala]: In your example, node B with a `null` `client` would try to instantiate it, so it would request the configuration to the driver and the next time it will already have a `client`.

So the configuration would be sent once per node as well as if it's broadcast. Do I missing something?

Sorry for the mess and thank you for help me to understand this `Spark` related 
things!

File: src/main/scala/tech/sourced/engine/util/Bblfsh.scala (hunk 4 above)
Comment (mcarmonaa) [reply src/main/scala/tech/sourced/engine/util/Bblfsh.scala]: Ok, let me change it! :+1:

File: src/main/scala/tech/sourced/engine/Engine.scala (hunk 6 above)
Comment (erizocosmico) [reply src/main/scala/tech/sourced/engine/Engine.scala]: `Engine` is not instantiated in all nodes, and thus, `setConfig` is not called in every node. Send does not mean, values will be available in every node if you store them somewhere, send means, I broadcast this value and give you a reference to it that is serializable, so in every task that will be sent to other nodes you will retrieve the same broadcasted value using that reference.
That's why the only why to access a broadcasted thing is having the broadcasted variable in scope.

File: src/main/scala/tech/sourced/engine/util/Bblfsh.scala [hunk 7]
```diff
@@ -1,50 +1,130 @@
...
     * Key used for the option to specify the host of the bblfsh grpc service.
     */
-  val hostKey = "spark.tech.sourced.bblfsh.grpc.host"
+  val bblfshHostKey = "spark.tech.sourced.bblfsh.grpc.host"
 
   /**
     * Key used for the option to specify the port of the bblfsh grpc service.
     */
-  val portKey = "spark.tech.sourced.bblfsh.grpc.port"
+  val bblfshPortKey = "spark.tech.sourced.bblfsh.grpc.port"
```
Comment (erizocosmico) [inline src/main/scala/tech/sourced/engine/util/Bblfsh.scala]: It's accessed as `Bblfsh.portKey`, the `bblfsh` prefix is kind of redundant given the scope of this property

**Comment by mcarmonaa** [comment]
@ajnavarro @erizocosmico thanks for the explanations!

Eventually, I moved the broadcast bblfsh configuration code to the `Bblfsh.getClient` method which is used from the udfs, so now the configuration should be broadcast properly I guess.

Is it correct?

File: src/main/scala/tech/sourced/engine/util/Bblfsh.scala [hunk 8]
```diff
@@ -1,50 +1,130 @@
...
+    }
+
+    config
+  }
+
+  private def getClient(): BblfshClient = synchronized {
     if (client == null) {
-      client = BblfshClient(config.host, config.port)
+      val Config(host, port) = getConfig()
+      if (broadcastConf) {
```
Comment (erizocosmico) [inline src/main/scala/tech/sourced/engine/util/Bblfsh.scala]: This still does not work on multiple nodes. You need to hold the reference to the `Broadcast` inside the serialized function that is going to be sent to the nodes to execute. This is only being changed in the node where `Engine` is initialized.

Comment (mcarmonaa) [reply src/main/scala/tech/sourced/engine/util/Bblfsh.scala]: I'll return again the config to a broadcast variable inside a closure where the udf is defined

**Comment by mcarmonaa** [comment]
Eventually, I back again to the approach of broadcast the bblfsh configuration variable from a closure.

I added to the UDF objects' `apply` method the session as an implicit parameter, so all of them can be used with `myCoolUDF(col1, col2)` making the `SparkSession` an implicit val in the scope.

File: src/main/scala/tech/sourced/engine/udf/ClassifyLanguagesUDF.scala [hunk 9]
```diff
@@ -1,34 +1,29 @@
...
-  /**
-    * Gets the language of the given file and returns the guessed language or none.
-    *
-    * @param isBinary whether it's a binary file or not
-    * @param path     file path
-    * @param content  file content
-    * @return `None` if no language could be guessed, `Some(language)` otherwise.
-    */
-  def getLanguage(isBinary: Boolean, path: String, content: Array[Byte]): Option[String] =
+  private def getLanguage(isBinary: Boolean, path: String, content: Array[Byte]): Option[String] = {
```
Comment (ajnavarro) [inline src/main/scala/tech/sourced/engine/udf/ClassifyLanguagesUDF.scala]: why did you remove existing documentation?

File: src/main/scala/tech/sourced/engine/udf/CustomUDF.scala [hunk 10]
```diff
@@ -3,13 +3,13 @@ package tech.sourced.engine.udf
 import org.apache.spark.sql.SparkSession
```
Comment (ajnavarro) [inline src/main/scala/tech/sourced/engine/udf/CustomUDF.scala]: This file should not be modified

File: src/main/scala/tech/sourced/engine/Engine.scala [hunk 11]
```diff
@@ -185,4 +185,5 @@ object Engine {
     new Engine(session)
       .setRepositoriesPath(repositoriesPath)
   }
+
```
Comment (ajnavarro) [inline src/main/scala/tech/sourced/engine/Engine.scala]: This file should not be modified.

File: src/main/scala/tech/sourced/engine/udf/ClassifyLanguagesUDF.scala (hunk 9 above)
Comment (mcarmonaa) [reply src/main/scala/tech/sourced/engine/udf/ClassifyLanguagesUDF.scala]: because now is private

Comment (ajnavarro) [reply src/main/scala/tech/sourced/engine/udf/ClassifyLanguagesUDF.scala]: Documentation is always good, so, if it was already there, I think is worth it keep there, even if the signature of the method is now private.

Comment (mcarmonaa) [reply src/main/scala/tech/sourced/engine/udf/ClassifyLanguagesUDF.scala]: Ok, I'll add it again, just keep in mind that documentation has to be maintained too :+1:

**Comment by mcarmonaa** [comment]
@ajnavarro can I merge this PR?

**Comment by ajnavarro** [comment]
LGTM
